pl_bolts/models/detection/yolo/yolo_module.py

```diff
@@ -0,0 +1,551 @@
+        classprobs, labels = torch.max(classprobs, -1)
+        boxes, confidences, classprobs, labels = self._filter_detections(boxes, confidences, classprobs, labels)
+        total_loss = torch.stack(tuple(losses.values())).sum()
+
+        for name, value in losses.items():
+            self.log('test/{}_loss'.format(name), value)
+        self.log('test/total_loss', total_loss)
+
+    def infer(self, image: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
+        """
+        Resizes given image to the network input size and feeds it to the network. Returns the
+        detected bounding boxes, confidences, and class labels.
+
+        Args:
+            image: An input image, a tensor of uint8 values sized ``[channels, height, width]``.
+
+        Returns:
+            boxes (:class:`~torch.Tensor`), confidences (:class:`~torch.Tensor`), labels (:class:`~torch.Tensor`):
+                A matrix of detected bounding box (x1, y1, x2, y2) coordinates, a vector of
+                confidences for the bounding box detections, and a vector of predicted class
+                labels.
+        """
+        network_input = image.float().div(255.0)
+        network_input = network_input.unsqueeze(0)
+        self.eval()
+        boxes, confidences, classprobs = self(network_input)
+        classprobs, labels = torch.max(classprobs, -1)
+        boxes, confidences, classprobs, labels = self._filter_detections(boxes, confidences, classprobs, labels)
+        assert len(boxes) == 1
+        boxes = boxes[0]
+        confidences = confidences[0]
+        labels = labels[0]
+
+        height = image.shape[1]
+        width = image.shape[2]
+        scale = torch.tensor([width, height, width, height], device=boxes.device)
+        boxes = boxes * scale
+        boxes = torch.round(boxes).int()
+        return boxes, confidences, labels
+
+    def load_darknet_weights(self, weight_file):
+        """
+        Loads weights to layer modules from a pretrained Darknet model.
+
+        One may want to continue training from the pretrained weights, on a dataset with a
+        different number of object categories. The number of kernels in the convolutional layers
+        just before each detection layer depends on the number of output classes. The Darknet
+        solution is to truncate the weight file and stop reading weights at the first incompatible
+        layer. For this reason the function silently leaves the rest of the layers unchanged, when
+        the weight file ends.
+
+        Args:
+            weight_file: A file object containing model weights in the Darknet binary format.
+        """
+        version = np.fromfile(weight_file, count=3, dtype=np.int32)
+        images_seen = np.fromfile(weight_file, count=1, dtype=np.int64)
+        print(
+            'Loading weights from Darknet model version {}.{}.{} that has been trained on {} '
+            'images.'.format(version[0], version[1], version[2], images_seen[0])
+        )
```

Comment:
@senarvi Is the `load_darnet_weights` functionality complete?

Reply:
It should be complete. I'm able to detect boxes using Darknet weights, and I'm able to continue training after loading Darknet weights. However, I got lower scores than when doing transfer learning using the Darknet implementation. I haven't got a chance to properly test it yet though, so I'm interested if you see any problems or if you're able to try it.

Reply:
I experimented with loading weights from a Darknet model. Training was very unstable, at least when using one of the truncated weight files. The Darknet weight files that are meant for transfer learning (such as [yolov4.conv.137](https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.conv.137)) are truncated up to the first detection layer, so that the number of object classes can be changed. The rest of the layers are initialized randomly. You can use them, but if you don't change the number of classes, you can load weights to all layers (such as [yolov4.weights](https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.weights)). I found that to work better. Otherwise the only way to make training stable was to use gradient clipping. It's strange because I think the Darknet implementation is not using gradient clipping. I spent some time trying to stabilize the operations such as divisions in IoU calculation, but to no avail.